Isc22 kl (#89)
easybuilders/easybuild-tutorial: docs/2022-isc22/integration_lumi.md

        capability using the Lustre parallel file system and Cray ClusterStor E1000 technology.
    -   The main disk based storage contains of 4 20 PB storage systems based on regular hard
        disks and also using the Lustre parallel file system.
    -   A 30 PB Ceph-based encrypted object storage system for storing, sharing and staging data
        will become available at a later date.

## Challenges

-   LUMI comes with the HPE Cray Programming Environment (PE) with the Cray Compiling Environment
    (a fully Clang/LLVM-based C/C++ compiler and a Fortran compiler using a Cray frontend but
    LLVM-based backend) and GNU and AMD compilers repackaged by Cray. The HPE Cray PE uses
    an MPICH-based MPI implementation (with libfabric backend on SlingShot 11) and also comes 
    with its own optimised mathematics libraries containing all the usual suspects you expect 
    from an EasyBuild toolchain. The HPE Cray PE environment is managed outside EasyBuild, but
    EasyBuild has a mechanism to integrate with it.

    Due to the specific hardware and software setup of LUMI using the EasyBuild common toolchains 

>     is all but straightforward. Getting Open MPI, a key component of the foss toolchain, to work 

Reviewer: ```suggestion
    is anything but straightforward. Getting Open MPI, a key component of the foss toolchain, to work 
```

    is a bit of a challenge at the moment. The Intel compilers are not a very good match with 
    AMD CPUs. One needs to be very careful when choosing the target architecture for code
    optimisation (`-xHost` will not generate AVX/AVX2 code), most recent versions of MKL have
    performance problems and sometimes produce wrong results and some recent MPI versions also
    cause problems with AMD CPUs.

-   LUMI has only a small central support team of 9 FTE. It is obvious that we cannot give the same
    level of support for software installations as some of the big sites such as JSC can do. Moreover,
    due to the nature of the LUMI environment with its Cray PE and novel AMD GPUs, installing software
    is more challenging than on your typical Intel + NVIDIA GPU cluster with NVIDIA/Mellanox

>     interconnect. Given that the technology is very new one can also expect a rapid evolution of

Reviewer: ```suggestion
    interconnect. Given that the technology is very new, one can also expect a rapid evolution of
```

    the programming environment with a risk of breaking compatibility with earlier versions,

>     and a need for frequent software updates in the initial yeas of operation to work around bugs

Reviewer: ```suggestion
    and a need for frequent software updates in the initial years of operation to work around bugs
```

-   The central support team may be small (the above 9 FTE), but the 
    LUMI consortium agreement states that the 
    consortium countries also have to assist in providing support. Given that neither the 
    LUMI User Support Team members nor the local support teams are employees of CSC this 
    also means that much of the application support has to be delivered with very little
    rights on the system (at most a working directory that is readable for all users). So we 
    need a setup where it is possible to help users installing applications while having
    only regular user rights on the system.

-   Due to the exploitation model and the small size of the central support team, license management
    is a pain. Users come to LUMI through 11 different channels (some with subchannels). Moreover it
    is the responsibility of the PI to invite users to a project and to ensure that they are eligible
    for LUMI use (taking into account, e.g., the European and USA export restrictions). At the central
    level we have no means currently to check who can use which software license. Hence we need a
    solution to distribute that responsibility also.

> -   I personally believe that users really want a customised software stack. They may say they want 

Reviewer: It will not be clear here who the "I" refers to. 
```suggestion
-   LUMI software support is based on the idea that users really want a customised software stack. They may say they want 
```

    exception are a few packages installed from binaries that are installed in a separate area across
    software stacks (e.g., ARM Forge and Vampir). The LUMI stack provides optimised binaries for each node
    type of LUMI, but some software that is not performance-critical is compiled only once. To this end we 
    have a partition corresponding to each node type, but also a common partition which is included with 
    the software of all other partitions. Software in that common partition can only have dependencies 
    in the common partition though.

    For now we keep the central LUMI software stack very small, but we provide an easy and fully transparent
    mechanism for the user to install software in their project directory that integrates fully with the LUMI
    stack. The user only needs to set an environment variable pointing to the project space.

In the future we envision that more software stacks may become available on LUMI as at least one local support
organisation wants to build their own stack. We also hope to find a solution
to get the foss toolchain working on at least the CPU nodes of LUMI with minimal changes

> in a future collaboration with HPE .

Reviewer: ```suggestion
in a future collaboration with HPE.
```

not only the names of the dependencies, is both a curse and a blessing. It is a curse when we need to upgrade
to a new compiler and also want to upgrade versions of certain dependencies, as a lot of easyconfig files need
to be checked and edited. In those cases the automatic concretiser of Spack may help to get running quicker.
But that very precise description is also a blessing when communicating with users as you can 
communicate with them through EasyBuild recipes (and possibly an easystack file, which defines a list of 
easyconfig files to install) rather than having a part of the specification in command line options of the
tool. So a user doesn't need to copy long command lines and as a support person
you know exactly what EasyBuild will do, so this leaves less room for errors and difficult to solve support
tickets. (Though note that with Spack environments you can also give a full specification in an Environment
manifest.)

### Offering EasyBuild in the LUMI stacks

The LUMI software stack is implemented as an Lmod hierarchy with two levels. On top of it EasyBuild is used
with a flat naming scheme (though we did consider a hierarchical one as well).

> The modules that load a specific version of the software stack are hand-written in LUA. The first level

Reviewer: ```suggestion
The modules that load a specific version of the software stack are hand-written in Lua. The first level
```

(a module called `LUMI`) is the version of the software stack, 

> the second level (a module called `oartition`) then choses the particular set of hardware to

Reviewer: ```suggestion
the second level (a module called `partition`) then choses the particular set of hardware to
```

symbolic links to the generic implementations, and so-called Lmod introspection functions are used by
the modules to determine their function from the location in the module tree. This makes it easy to correct
bugs and extend the module files without having to go over different implementations in different locations.

To enable all this functionality, including the common partition, a double module structure was needed
to ensure the proper working of the Lmod hierarchy. The infrastructure tree is implemented following
all proper rules for an Lmod hierarchy. It houses the `LUMI` and `partition` modules among others, but
also the toolchain modules so that they can be configured specifically for each version of the `LUMI` 
software stack and `partition`. The software module tree contains all software installed through 
EasyBuild, with room to expand to manually installed software or software installed with Spack
should this turn out to be necessary and safe to combine with EasyBuild-managed software.

EasyBuild is configured through some configuration modules that set a range of `EASYBUILD_*` variables

> to configure EasyBuild. There is again only a single EasyBuild configuration module implemented in LUA,

Reviewer: ```suggestion
to configure EasyBuild. There is again only a single EasyBuild configuration module implemented in Lua,
```

remains consistent. This is important, e.g., to guarantee that the user configuration builds correctly
upon the centrally installed stack, so any change to the structure of the latter should also propagate
to the configuration of the former.

## EasyBuild on LUMI

### Custom toolchains

The LUMI EasyBuild solution is not based on the common toolchains as the HPE Cray Programming Environment
is the primary development environment on LUMI for which we also obtain support from HPE and as the common
toolchains both post problems on LUMI. So far we have not yet succeeded to ensure proper operation of 
Open MPI on LUMI which is a showstopper for working with the FOSS toolchain hierarchy. There is also no support
yet for AMD GPUs in the common toolchains. Moreover, the FOSS toolchain is becoming a very powerful vehicle
by the inclusion of FlexiBLAS and now also efforts to build an Open MPI configuration that supports both
CPU nodes and NVIDIA GPU nodes but the downside of this is that it has also become a very complicated 
setup to adapt if you need something else, as we do on LUMI. The Intel toolchain also has its problems. 

> Intel does not support oneAPI on AMD processors. Though Intel MPI can in principle run on SlingShot,
> also the SlingShot-11 update that is being installed on LUMI right now during this tutorial, as it is

Reviewer: ```suggestion
Intel does not support oneAPI on AMD processors. Though Intel MPI can in principle run on SlingShot as it is
```
Drop something that will be out  of date really soon
